File: solutions/blind-75/string/groupAnagrams.js
const ALPHABET_SIZE = 26;
const FIRST_CHAR = 'a';

const createArrayOfSize = (size, fillValue) => {
  return new Array(size).fill(fillValue);
};

const getAsciiValue = (char) => char.charCodeAt(0);

const getIndex = (char) => getAsciiValue(char) - getAsciiValue(FIRST_CHAR);

const getCharFrequencies = (string) => {
  const charFrequencies = createArrayOfSize(ALPHABET_SIZE, 0);

  string.split('').forEach((char) => {
    const index = getIndex(char);

    charFrequencies[index] += 1;
  });

  return JSON.stringify(charFrequencies);
};

const groupAnagrams = (strs) => {
  const anagramGroupings = {};

  strs.forEach((string) => {
    const charFrequencies = getCharFrequencies(string);

    const charFrequenciesInAnagramGroupings =
      anagramGroupings.hasOwnProperty(charFrequencies);

    if (!charFrequenciesInAnagramGroupings)
      anagramGroupings[charFrequencies] = [];

    anagramGroupings[charFrequencies].push(string);
  });

  return Object.values(anagramGroupings);
};

File: solutions/python/general-algos/strings/group_anagrams.py
from collections import defaultdict

ALPHABET_SIZE = 26
LOWERCASE_A = 'a'

class Solution:
    def groupAnagrams(self, strings: List[str]) -> List[List[str]]:
        anagram_groupings = defaultdict(list)
        
        for string in strings:
            char_frequencies = self.get_char_frequencies(string)
            anagram_groupings[char_frequencies].append(string)
        
        return list(anagram_groupings.values())
    
    def get_char_frequencies(self, string):
        char_frequencies = [0 for _ in range(ALPHABET_SIZE)]
        
        for char in string:
            index = ord(char) - ord(LOWERCASE_A)
            char_frequencies[index] += 1
        
        return tuple(char_frequencies)